### ISSUE TITLE
Tags rather than names in variable parts of the metrics

### ISSUE THREAD

**Issue by potiuk** [apache/airflow#11463]
**Description**

It would be great if the metrics we publish in Airflow change to include TAGs for "variable" parts of those (such as dag_id, task_id, pool name) rather than being part of the metrics.

See here: https://airflow.apache.org/docs/1.10.12/metrics.html, https://airflow.readthedocs.io/en/latest/logging-monitoring/metrics.html  - those variable parts should be tags.

We might consider to change those metrics by default in 2.0 and possibly introduce a backport in 1.10.13 to allow people to migrate.

**Use case / motivation**

Having the variable parts of the metrics makes it really difficult to aggregate the metrics. Seems that statsd python library does not support tags, but there are possible solutions by installing extensions:

https://stackoverflow.com/questions/49852654/sending-statsd-metrics-with-tags-from-python-client

**Related Issues**

**Comment by potiuk** [comment]
@ashb  @kaxil  -> you have an experience with running Airflow in Production environment with Graphana dashboards etc. Do you think this one is a valid and important one ? Do you think we should target it for 2.0 and possibly backport to 1.10 for easier migration? This seems like an easy task, to not remove existing metrics but add new ones with tags instead.

**Comment by ashb** [comment]
Plain statsd doesn't support "tags" is the main issue here

**Comment by ashb** [comment]
https://statsd.readthedocs.io/en/latest/tags.html?highlight=tag

> Tagged metrics—such as those used by Datadog and Telegraf—are explicitly outside the scope of this library. Alternatives exist and are recommended. This document lays out the reasons to avoid support for tags.

Yeah, so tagged metrics are not in main pystatsd client, not in "stock" statsd server.

Supporting them would be good, but we possibly need to be slightly smarter about how we do it (such as convert tags in to names for fallback? Not sure)

**Comment by kaxil** [comment]
It would be better to natively support Prometheus (https://github.com/prometheus/client_python)

**Comment by ashb** [comment]
Native prom support would be good - but that needs a) a webserver, and b) some way of collecting metrics from across the different processes, and for components without webservers (scheduler and workers right now)

**Comment by malthe** [comment]
In addition to these tags, the actual DAG `tags` should also be provided as part of this.

The following docs do seem to suggest that tags are supported in the StatsD format: https://docs.aws.amazon.com/AmazonCloudWatch/latest/monitoring/CloudWatch-Agent-custom-metrics-statsd.html.
```
MetricName:value|type|@sample_rate|#tag1:
  value,tag1...
```

**Comment by syun64** [comment]
Hi folks - we are actively interested in working on this item as users of Airflow x Graphana dashboards

The current state of airflow-stats module without any tagging capability is very limiting for us. Here are some example use cases where introducing tagging would be a significant improvement.
1.  Setting up Graphana dashboards to track and aggregate similar metrics (e.g. "dag.<dag_id>.<task_id>.duration") is difficult, as each new task or DAG results in a new name. If we use dag.duration with tags like dag_id="dag_1", task_id = "task_x" instead, our metric aggregation will be greatly simplified on Graphana
2. Setting up alerting rules against similar statsd metrics (e.g. "operator_failures_<operator_name>") is difficult as each new task or DAG results in a new name. If we were to publish this metric as "operator_failure" with tag operator_name = "operator_y", we will be able to set up alerting rules against the metric "operator_failure"

As users of [influxdb-statsd](https://www.influxdata.com/blog/getting-started-with-sending-statsd-metrics-to-telegraf-influxdb/), we are interested in introducing an optional config-driven change that will allow us add these input variable values as tags rather than as concatenated names.

Keeping in mind that different implementations of Statsd has different ways of tagging, I think it would be necessary to have different implementations of the tagging interface for different supported statsd plugins. Similar to how we instantiate the [SafeDogStatsdLogger](https://github.com/apache/airflow/blob/main/airflow/stats.py#L294) if we have the option "statsd_datadog_enabled" set to True, we could instantiate the SafeInfluxDBStatsdLogger that implements the specified tagging approach if statsd_influxdb_enabled is set to True.

As a part of this exercise, I'm wondering if revising the current [StatsLogger](https://github.com/apache/airflow/blob/main/airflow/stats.py#L53) Protocol to have "tags" parameter as an OrderedDict in its method signatures would be a good idea. Just as it has already been proposed before on https://github.com/apache/airflow/pull/12158 we could then use these **input tags** to either just publish the tags to tag-supported plugin, or concatenate the tags in the given order at the end of the stats name.

```
class StatsLogger(Protocol):
    @classmethod
    def incr(cls, stat: str, count: int = 1, rate: int = 1, tags: OrderedDict[str,Union[str,int]]) -> None:
        """Increment stat."""
...

```

```
def concatenate_tags_to_classic_stat(fn: T) -> T:
    """
    Concatenate tags to statname if using classic Statsd
    """

    @wraps(fn)
    def wrapper(_self, stat=None, tags=OrderedDict(), *args, **kwargs):
        if stat is not None:
            for k, v in tags.items():
                stat += f".{v}"
        return fn(_self, stat, *args, **kwargs)

    return cast(T, wrapper)

class SafeStatsdLogger:
    @concatenate_tags_to_classic_stat
    @validate_stat
    def incr(self, stat, count=1, rate=1):
        """Increment stat."""
        if self.allow_list_validator.test(stat):
            return self.statsd.incr(stat, count, rate)
        return None
...

```

```
class SafeDogStatsdLogger:
    @validate_stat
    def incr(self, stat, count=1, rate=1, tags=None):
        """Increment stat."""
        if self.allow_list_validator.test(stat):
            tags = [f"{k}:{v}" for k, v in tags.items()] if tags else None
            return self.dogstatsd.increment(metric=stat, value=count, tags=tags, sample_rate=rate)
        return None
...

```

```
def concatenate_tags_to_influxdb_statname(fn: T) -> T:
    """
    Concatenate tags to statname as InfluxDB recognized Statsd tags
    """

    @wraps(fn)
    def wrapper(_self, stat=None, tags=OrderedDict(), *args, **kwargs):
        if stat is not None:
            for k, v in tags.items():
                stat += f",{k}={v}"
        return fn(_self, stat, *args, **kwargs)

    return cast(T, wrapper)

class SafeInfluxDBStatsdLogger:
    @concatenate_tags_to_influxdb_statname
    @validate_stat
    def incr(self, stat, count=1, rate=1, tags=None):
        """Increment stat."""
        if self.allow_list_validator.test(stat):
            return self.statsd.incr(stat, count, rate)
        return None
...

```

There are some open questions regarding how this proposal would be a breaking change for some of the existing metrics like: "dag.<dag_id>.<task_id>.duration", and whether or not we will support the addition of new tags to existing metrics as breaking changes to Classic Statsd. However, I think supporting tags on statsd plugins and normalizing the tag publishing interface is an improvement worth the effort to revisit this conversation. Please let me know what your thoughts are on this proposal, and whether you'd be open to me opening a PR and organizing some follow up discussions to formalize the proposal.
@potiuk @ashb

**Comment by potiuk** [comment]
I think it would be better if you focus on implementing https://cwiki.apache.org/confluence/display/AIRFLOW/AIP-49+OpenTelemetry+Support+for+Apache+Airflow 

Implementing Open Telemetry has alrready been designed, discussed and approved and also POC was developed. 

It needs helping hands to implement it.  Many of the questions you discuss before has been answered in the Open Telemetry AIP and implementing Open Telemetry is WAAAAAY better approach rather than modifying existing statsd metrics - especially that it will work in the same way for multiple telemetry solutions.

cc: @howardyoo

**Comment by syun64** [comment]
Thank you for the quick response @potiuk ! I'm starting to really appreciate the friendliness and responsiveness of this community, and it really shows with every single response I get!

I do agree that the idea of Open Telemetry implementation makes a whole lot of sense, especially from the perspective of maintaining Airflow in a stable manner by abiding to a well-defined tagging standard. 

Unfortunately, I'm not sure if we will be able to adopt Airflow-OpenTelemetry Metrics out of the box as easily as we think. The main hurdles for some folks in big companies like us will just be that we rely on managed infrastructures that only accept certain protocols into their metric storage. In our case, we are only able to accept Statsd/influxdb-statsd UDP protocols which is then aggregated and propagated to our MetricTank down the pipeline. I'm not an expert on the exact factors that ultimately drove our managed infrastructure to hold off on adopting OTLP as a supported input protocol at the company-wide level, but I'm guessing it would have to do with the fact that gRPC/HTTP introduces more network traffic compared to UDP - to a factor that might have a significant performance impact if adopted at scale.

Another idea I had to try to adopt Airflow-OpenTelemetry with our existing infrastructure was to use a StatsdExporter - but this seems to be [blocked by OpenTelemetry community](https://github.com/open-telemetry/opentelemetry-specification/issues/374) for reasons we all very well know.

Given these limitations, I think it would make more sense to introduce callbacks that emit Statsd metrics in the dag definitions themselves if we aren't able to introduce the tagging related changes above... Adopting Airflow-OpenTelemetry just doesn't seem possible for us just yet.

**Comment by potiuk** [comment]
Yeah. Given the context (thanks for that) I think small - non-breaking especially - improvements in this area are still good tactical solution even if the long-term strategic approach is with OpenTelemetry. The Open-Telemetry is long-time approved and it did not have enough of a priority to be implemented, so it might well be a long time until it gets implemented and released.

**Comment by howardyoo** [comment]
Congrats to the close of a long lasted git issue! :-)
